### PR TITLE
COL-455 COL-668 Generate preview and thumbnail images for exported whiteboards

### PR DESCRIPTION
**Pull request by paulkerschen**
https://jira.ets.berkeley.edu/jira/browse/COL-455
and
https://jira.ets.berkeley.edu/jira/browse/COL-668

have occult connections.

### CODE REVIEW

File: node_modules/col-assets/lib/api.js [hunk 1]
```diff
@@ -1035,13 +1038,13 @@ var generatePreviews = module.exports.generatePreviews = function(asset, callbac
...
 
   var assetUrl = null;
-  if (asset.type === 'file') {
+  if (asset.type === 'file' || asset.type === 'whiteboard') {
     assetUrl = asset.download_url;
   } else if (asset.type === 'link') {
     assetUrl = asset.url;
-  } else {
-    // Don't generate previews for whiteboards, the Whiteboards API
-    // takes care of this
```
Comment (paulkerschen) [inline node_modules/col-assets/lib/api.js]: Part 1 of the bug: the assets API assumed that the whiteboards API would handle previews for exported whiteboards.

File: node_modules/col-whiteboards/lib/api.js [hunk 2]
```diff
@@ -1439,38 +1439,25 @@ var exportWhiteboardToAsset = module.exports.exportWhiteboardToAsset = function(
...
 
-        generateThumbnail(whiteboard, function(err, whiteboard, dimensions) {
+        AssetsAPI.createWhiteboard(ctx, whiteboard, title, opts, function(err, asset) {
           if (err) {
-            return log.error({'err': err}, 'Could not generate a thumbnail for an exported whiteboard');
+            return callback(err);
           }
 
-          if (whiteboard.image_url) {
-            updateAssetPreview(asset.id, whiteboard.image_url, dimensions.width, function(err) {
```
Comment (paulkerschen) [inline node_modules/col-whiteboards/lib/api.js]: Part 2 of the bug: the whiteboards API was not calling `AssetsAPI.updateAssetPreview` (which would have performed properly) but `WhiteboardsAPI.updateAssetPreview` (which does something completely different).

With this refactor, `AssetsAPI.createWhiteboard` (above) now calls `AssetsAPI.updateAssetPreview` as part of its business.

File: node_modules/col-assets/tests/test-assets.js [hunk 3]
```diff
@@ -1047,24 +1047,26 @@ describe('Assets', function() {
...
-                                          // Verify that the creator cannot delete the asset
-                                          AssetsTestUtil.assertDeleteAssetFails(client1, course, asset4.id, 401, function() {
-
-                                            // Delete the exported whiteboard
-                                            AssetsTestUtil.assertDeleteAsset(client1, course, exportedAsset.id, function() {
-                                              // Verify that the creator can delete the asset
-                                              AssetsTestUtil.assertDeleteAsset(client1, course, asset4.id, function() {
-
-                                                return callback();
+                              AssetsTestUtil.mockPreviewData(asset4, function() {
```
Comment (paulkerschen) [inline node_modules/col-assets/tests/test-assets.js]: This added line is the only change in the test; the rest is indentation.

File: node_modules/col-whiteboards/lib/api.js [hunk 4]
```diff
@@ -1439,38 +1439,25 @@ var exportWhiteboardToAsset = module.exports.exportWhiteboardToAsset = function(
...
-
-        // Pass the asset to the caller
-        callback(null, asset, whiteboard);
-
-        // Asynchronously create a thumbnail for the exported whiteboard asset. If the preview integration
-        // has not been enabled, we can't generate a thumbnail and simply return
-        if (!config.get('previews.enabled')) {
-          return;
+          log.error({'err': err}, 'Error creating a PNG from a newly exported whiteboard');
+          return callback({'code': 500, 'msg': 'The board could not be exported'});
```
Comment (johncrossman) [inline node_modules/col-whiteboards/lib/api.js]: @paulkerschen is `whiteboard.id` available to put in `log.error()`?

**Comment by johncrossman** [comment]
Looks good to me, @paulkerschen  Put one question

File: node_modules/col-whiteboards/lib/api.js (hunk 4 above)
Comment (paulkerschen) [inline node_modules/col-whiteboards/lib/api.js]: Available and added, @johncrossman, thanks.

**Pull request merged**